src/main/scala/org/apache/spark/ml/ModelBuilder.scala

```diff
@@ -19,4 +24,193 @@ object ModelBuilder {
       coefficients: Vector,
       intercept: Double): GeneralizedLinearRegressionModel =
     new GeneralizedLinearRegressionModel("glr-uid", coefficients, intercept)
-}
\ No newline at end of file
+
+  def newDecisionTreeClassificationModel(
+      depth: Int,
+      numClasses: Int,
+      featureArity: Array[Int],
+      seed: Long): DecisionTreeClassificationModel = {
+    require(numClasses >= 2, s"DecisionTreeClassificationModel requires numClasses >= 2," +
+      s" but was given $numClasses")
+    val rootNode = TreeBuilder.randomBalancedDecisionTree(depth = depth, labelType = numClasses,
+      featureArity = featureArity, seed = seed)
+    new DecisionTreeClassificationModel(rootNode, numFeatures = featureArity.length,
+      numClasses = numClasses)
+  }
+
+  def newDecisionTreeRegressionModel(
+      depth: Int,
+      featureArity: Array[Int],
+      seed: Long): DecisionTreeRegressionModel = {
+    val rootNode = TreeBuilder.randomBalancedDecisionTree(depth = depth, labelType = 0,
+      featureArity = featureArity, seed = seed)
+    new DecisionTreeRegressionModel(rootNode, numFeatures = featureArity.length)
+  }
+}
+
+/**
+ * Helpers for creating random decision trees.
+ */
+object TreeBuilder {
+
+  /**
+   * Generator for a pair of distinct class labels from the set {0,...,numClasses-1}.
+   * Pairs are useful for trees to make sure sibling leaf nodes make different predictions.
+   * @param numClasses  Number of classes.
+   */
+  private class ClassLabelPairGenerator(val numClasses: Int)
+    extends RandomDataGenerator[Pair[Double, Double]] {
+
+    require(numClasses >= 2,
+      s"ClassLabelPairGenerator given label numClasses = $numClasses, but numClasses should be >= 2.")
+
+    private val rng = new java.util.Random()
+
+    override def nextValue(): Pair[Double, Double] = {
+      val left = rng.nextInt(numClasses)
+      var right = rng.nextInt(numClasses)
+      while (right == left) {
+        right = rng.nextInt(numClasses)
+      }
+      new Pair[Double, Double](left, right)
+    }
+
+    override def setSeed(seed: Long) {
+      rng.setSeed(seed)
+    }
+
+    override def copy(): ClassLabelPairGenerator = new ClassLabelPairGenerator(numClasses)
+  }
+
+
+  /**
+   * Generator for a pair of real-valued labels.
+   * Pairs are useful for trees to make sure sibling leaf nodes make different predictions.
+   */
+  private class RealLabelPairGenerator() extends RandomDataGenerator[Pair[Double, Double]] {
```

Comment:
can you pass a seed to the `rng` below? You already do that in `randomBalancedDecisionTree`